AdminServer/appscale/admin/utils.py

```diff
@@ -497,3 +503,119 @@ def queues_from_dict(payload):
     queues[name] = queue
 
   return {'queue': queues}
+
+
+def retry_coroutine(func, async=True, backoff_base=2, backoff_multiplier=0.2,
+                    backoff_threshold=300, max_retries=None,
+                    retry_on_exception=None, refresh_args_kwargs_func=None):
+  """ Wraps func with retry mechanism which runs up to max_retries attempts
+  with exponential backoff (sleep = backoff_multiplier * backoff_base**X).
+  
+  Args:
+    func: function or coroutine to wrap.
+    async: a boolean indicating if result of function should be yielded.
+    backoff_base: a number to use in backoff calculation.
+    backoff_multiplier: a number indicating initial backoff.
+    backoff_threshold: a number indicating maximum backoff.
+    max_retries: an integer indicating maximum number of retries.
+    retry_on_exception: a function receiving one argument: exception object and
+      returning True if retry makes sense for such exception.
+    refresh_args_kwargs_func: a function without arguments which should return
+      tuple of two elements (arguments list, keyword arguments dict).
+  Returns:
+    A tornado coroutine wrapping function with retry mechanism.
+  """
+
+  @gen.coroutine
+  def persistent_execute(*args, **kwargs):
+    retries = 0
+    backoff = backoff_multiplier * (backoff_base ** 1)
+
+    while True:
+      try:
+        if retries and refresh_args_kwargs_func is not None:
```

Comment:
I don't think this refresh should be needed. If the watched node gets updated during the first execution of the function, the function will be triggered again when the watch is re-established. Since this function yields, though, you may want to use a tornado async lock to prevent the second trigger from being run before the first finishes.

Reply:
I was trying to resolve following problem:
 0. AdminServer known_projects=[projA, projB, projC]
 1. children are updated, new list is [projA, projB]
 2. watch finds out that projC should be removed, but fails to do this
 3. retrying after 0.2s, 0.4s, 0.8s, 1.6s, 3.2s, next retry in 6.4s
 4. children are updated, new list is [projA, projB, projD]
 5. update is handled properly, projC is removed, projD is created. Now known_projects=[projA, projB, projD]
 6. retry num# 6 with [projA, projB] removes projD

Reply:
That is definitely the right situation to consider. An async lock will prevent this by serializing the triggers. One problem with this implementation is that it adds another call that can fail or block the loop. While it makes race conditions less likely, it doesn't completely resolve the situation you describe. For example,

1. Trigger 1 (T1) gets scheduled to run on IOLoop (projects A, B)
2. T1 gets the current list (A, B)
3. The children are updated (A, B, D)
4. The ZK thread schedules trigger 2 to run (A, B, D)
5. T1 tries to yield result from function 
6. Function from T1 gets scheduled to run on IOLoop with args (A, B) (F1)
7. T2 gets the current list (A, B, D)
8. T2 tries to yield result from function
9. Function from T2 gets scheduled to run with args (A, B, D) (F2)

Now it's a race between F1 and F2. If any async operation in F1 (an http call to monit, for example) takes a while, F2 could finish before F1.

Reply:
I see. As you suggest i need to wrap watch functions with async lock to avoid concurrent update. But async sleep between retries must be outside lock (to be reactive to updates from zk), so i need to make sure i don't overwrite latest update. I'll try to find the best solution) thanks for raising this issue!

Reply:
Okay, it's up to you how to implement it, but I personally would be fine with the sleep inside the lock. The reactivity comes from the watch, which will ensure that the function gets called with the latest state. Retrying the function with slightly stale state is not a problem because it will get called again with fresh state as soon as ZK is available again.